### PR TITLE
MISUV-7805-one-year-with-q-updates

### PR DESCRIPTION
**Pull request by james-king-hmrc**
one-year-with-q-updates

### CODE REVIEW

File: app/services/optout/OptOutService.scala [hunk 1]
```diff
@@ -185,33 +198,33 @@ class OptOutService @Inject()(itsaStatusUpdateConnector: ITSAStatusUpdateConnect
...
 
-  def getSubmissionCountForTaxYear(offeredOptOutYears: Seq[TaxYear])
-                                  (implicit user: MtdItUser[_], hc: HeaderCarrier, ec: ExecutionContext): Future[SubmissionsCountForTaxYearModel] = {
-    val futureCounts: Seq[Future[SubmissionsCountForTaxYear]] = offeredOptOutYears.map { optOutTaxYear =>
-      nextUpdatesService.getSubmissionCounts(optOutTaxYear)
+  def getQuarterlyUpdatesCountForTaxYear(offeredOptOutYears: Seq[TaxYear])
+                                        (implicit user: MtdItUser[_], hc: HeaderCarrier, ec: ExecutionContext): Future[QuarterlyUpdatesCountForTaxYearModel] = {
+    val futureCounts: Seq[Future[QuarterlyUpdatesCountForTaxYear]] = offeredOptOutYears.map { optOutTaxYear =>
+      nextUpdatesService.getQuarterlyUpdatesCounts(optOutTaxYear)
     }
```
Comment (SiddharthAtos) [inline app/services/optout/OptOutService.scala:205]: should fix: this part can be done with 1 api call, combining tax year.  using - 
```scala
ObligationsConnector.getAllObligations(fromDate: LocalDate, toDate: LocalDate)
```

File: app/services/NextUpdatesService.scala [hunk 2]
```diff
@@ -98,12 +98,12 @@ class NextUpdatesService @Inject()(val obligationsConnector: ObligationsConnecto
 
   }
 
-  def getSubmissionCounts(optOutTaxYear: TaxYear)
-                         (implicit hc: HeaderCarrier, mtdUser: MtdItUser[_]): Future[SubmissionsCountForTaxYear]  = {
+  def getQuarterlyUpdatesCounts(optOutTaxYear: TaxYear)
```
Comment (SiddharthAtos) [inline app/services/NextUpdatesService.scala:101]: should fix: `optOutTaxYear` doesn't give any context of `NextUpdatesService`, could rename it to `taxYear`

File: app/services/optout/OptOutService.scala [hunk 3]
```diff
@@ -157,15 +157,28 @@ class OptOutService @Inject()(itsaStatusUpdateConnector: ITSAStatusUpdateConnect
...
+      } getOrElse Future.successful(noQuarterlyUpdates)
     }
+
+    for {
+      proposition <- fetchOptOutProposition()
+      intent <- fetchSavedIntent()
+      propositionType <- Future.successful(proposition.optOutPropositionType)
+      quarterlyUpdatesCount <- getQuarterlyUpdatesCount(propositionType)
+      viewModel <- Future.successful(processPropositionType(propositionType.get, intent, quarterlyUpdatesCount))
+    } yield viewModel
```
Comment (SiddharthAtos) [inline app/services/optout/OptOutService.scala:181]: Should fix: Lets simplify this 
```scala
for {
  proposition <- fetchOptOutProposition()
  intent <- fetchSavedIntent()
  propositionType = proposition.optOutPropositionType
  quarterlyUpdatesCount <- getQuarterlyUpdatesCount(propositionType)
} yield processPropositionType(propositionType.get, intent, quarterlyUpdatesCount)
```

File: app/services/optout/OptOutService.scala [hunk 4]
```diff
@@ -157,15 +157,28 @@ class OptOutService @Inject()(itsaStatusUpdateConnector: ITSAStatusUpdateConnect
...
+        case p: OneYearOptOutProposition => Some(OneYearOptOutCheckpointViewModel(intent = p.intent.taxYear, state = p.state(), Some(quarterlyUpdatesCount)))
+        case _: MultiYearOptOutProposition => intent.map(i => MultiYearOptOutCheckpointViewModel(intent = i))
+    }
+
+    def getQuarterlyUpdatesCount(propositionType: Option[OptOutPropositionTypes]): Future[Int] = {
+      propositionType.map {
+        case p: OneYearOptOutProposition =>
+          getQuarterlyUpdatesCountForTaxYear(Seq(p.intent.taxYear)).map(v => v.getCountFor(p.intent.taxYear))
+        case _ => Future.successful(noQuarterlyUpdates)
+      } getOrElse Future.successful(noQuarterlyUpdates)
```
[the remark this answers is not in the document]
Comment (james-king-hmrc) [reply app/services/optout/OptOutService.scala]: done

File: app/services/optout/OptOutService.scala (hunk 3 above)
Comment (james-king-hmrc) [reply app/services/optout/OptOutService.scala]: done

File: app/services/NextUpdatesService.scala (hunk 2 above)
Comment (james-king-hmrc) [reply app/services/NextUpdatesService.scala]: done

File: app/services/optout/OptOutService.scala [hunk 5]
```diff
@@ -185,33 +198,33 @@ class OptOutService @Inject()(itsaStatusUpdateConnector: ITSAStatusUpdateConnect
...
+  private val noQuarterlyUpdates = 0
 
   private def includeTaxYearCount(optOutYear: TaxYear)(countsYear: TaxYear): Boolean = {
     (optOutYear.startYear == countsYear.startYear) || (optOutYear.endYear == countsYear.startYear)
   }
 
-  case class SubmissionsCountForTaxYearModel(counts: Seq[SubmissionsCountForTaxYear]) {
+  case class QuarterlyUpdatesCountForTaxYearModel(counts: Seq[QuarterlyUpdatesCountForTaxYear]) {
 
     def getCountFor(optOutYear: TaxYear): Int = counts.find(v => includeTaxYearCount(optOutYear)(v.taxYear))
```
Comment (SiddharthAtos) [inline app/services/optout/OptOutService.scala:221]: should fix: don't need this
```scala
private def includeTaxYearCount(optOutYear: TaxYear)(countsYear: TaxYear): Boolean = {
    (optOutYear.startYear == countsYear.startYear) || (optOutYear.endYear == countsYear.startYear)
  }
```
It is similar to 
``` scala
find(optOutTaxYear)
```

File: test/views/optOut/ConfirmOptOutViewSpec.scala [hunk 6]
```diff
@@ -90,5 +90,9 @@ class ConfirmOptOutViewSpec extends TestSupport {
       pageDocument.getElementById("cancel-button").text() shouldBe confirmOptOutMessages.cancelButton
       pageDocument.getElementById("confirm-optout-form").attr("action") shouldBe confirmOptOutMessages.confirmedOptOutURLAgent
     }
+
+    "with quarterly updates as zero count" in new Setup(false) {
+      pageDocument.getElementById("warning-inset") shouldBe null
```
Comment (SiddharthAtos) [inline test/views/optOut/ConfirmOptOutViewSpec.scala:95]: should fix: lets not use `null`
substitute it with
```scala
pageDocument.select("#warning-inset").size() shouldBe 0
```

File: test/views/optOut/ConfirmOptOutViewWithQuarterlyUpdatesSpec.scala
Comment (SiddharthAtos) [inline test/views/optOut/ConfirmOptOutViewWithQuarterlyUpdatesSpec.scala:1]: should fix: Why create new spec file for small addition. Can't we add same to `test/views/optOut/ConfirmOptOutViewSpec.scala` as its part of same view?

reason: it adds difficulties while maintaining two spec for same view.

Comment (james-king-hmrc) [reply test/views/optOut/ConfirmOptOutViewWithQuarterlyUpdatesSpec.scala:1]: done

File: test/views/optOut/ConfirmOptOutViewSpec.scala (hunk 6 above)
Comment (james-king-hmrc) [reply test/views/optOut/ConfirmOptOutViewSpec.scala]: done

**Review by SiddharthAtos — APPROVED** [review]

**Pull request merged**